### PR TITLE
FIX: Better infinite scrolling on categories page

### PR DESCRIPTION
**Pull request by nbianca**
This commit refactor CategoryList to remove usage of EmberObject,
hopefully make the code more readable and fixes various edge cases with
lazy loaded categories (third level subcategories not being visible,
subcategories not being visible on category page, requesting for more
pages even if the last one did not return any results, etc).

The problems have always been here, but were not visible because a lot
of the processing was handled by the server and then the result was
serialized. With more of these being moved to the client side for the
lazy category loading, the problems became more obvious.

### CODE REVIEW

File: app/assets/javascripts/discourse/app/models/category-list.js [hunk 1]
```diff
@@ -9,83 +8,79 @@ import Topic from "discourse/models/topic";
...
     const newCategoryList = CategoryList.categoriesFrom(this.store, result);
-    this.categories.pushObjects(newCategoryList.categories);
+    newCategoryList.forEach((c) => this.categories.pushObject(c));
   },
 });
 
 CategoryList.reopenClass({
-  categoriesFrom(store, result) {
-    const categories = CategoryList.create({ store });
-    const list = Category.list();
```
Comment (nbianca) [inline app/assets/javascripts/discourse/app/models/category-list.js:55]: There is no reason to pass `list` as an argument to the other functions when it is globally accessible.

File: app/models/category_list.rb [hunk 2]
```diff
@@ -156,7 +156,11 @@ def find_categories
     notification_levels = CategoryUser.notification_levels_for(@guardian.user)
     default_notification_level = CategoryUser.default_notification_level
 
-    if @options[:parent_category_id].blank?
+    if SiteSetting.lazy_load_categories
+      @categories.each do |c|
+        c.subcategory_ids = Category.where(parent_category_id: c.id).pluck(:id)
```
Comment (tgxworld) [inline app/models/category_list.rb:161]: This is going to cause an N+1 queries problem because a DB query is being triggered for each iteration in the loop.

File: app/models/category_list.rb [hunk 3]
```diff
@@ -156,7 +156,11 @@ def find_categories
...
+      end
```
Comment (tgxworld) [inline app/models/category_list.rb:162]: Do we need to test the changes here?

**Comment by tgxworld** [comment]
> fixes various edge cases with
lazy loaded categories (third level subcategories not being visible,
subcategories not being visible on category page, requesting for more
pages even if the last one did not return any results, etc).

Do you think we should start adding tests for problems mentioned above? My thinking here is that the `lazy_load_categories` flag is eventually going to be remove and the new behaviour made the default so we should start writing tests to prevent the possible regressions in the future.

File: app/models/category_list.rb (hunk 3 above)
Comment (nbianca) [reply app/models/category_list.rb:162]: I added a test for this.

File: app/models/category_list.rb (hunk 2 above)
Comment (nbianca) [reply app/models/category_list.rb:161]: Oh yes. I implemented this quickly for the prototype and forgot about it. I fixed it.

**Comment by nbianca** [comment]
It the next iteration, I plan to add more tests on both the backend and the frontend. We're going to refactor some more things anyway.

**Review by tgxworld — APPROVED** [review]
